cabal-install/Distribution/Client/List.hs

```diff
@@ -129,6 +138,94 @@ getPkgList verbosity packageDBs repoCtxt comp conf listFlags pats = do
       , pkg <- search index pat ]
 
 
+-- | List the installed packages that the current package depends on.
+-- This is similar to getPkgList, but it takes no patterns; rather, it assumes
+-- there is a current, configured configuration so it can determine the exact
+-- (installed) packages depended on.
+listPkgDeps :: Verbosity
+            -> RepoContext
+            -> ListFlags
+            -> IO ()
+listPkgDeps verbosity repoCtxt listFlags =
+    case possiblyShowPackageFields of
+      Left msg                -> die msg
+      Right showPackageFields -> do
+          installedPkgIndex <-
+            fmap Distribution.Simple.LocalBuildInfo.installedPkgs $
+              getPersistBuildConfig (useDistPref defaultSetupScriptOptions)
+               -- TODO: add a --builddir=DIR option rather than assume
+               -- the default value of 'dist'
+
+          srcPkgIndex <- fmap packageIndex $
+                           getSourcePackages verbosity repoCtxt
+
+          let
+              showPkgSimple pkg =
+                display (packageName pkg)
+                ++ " "
+                ++ display (packageVersion pkg)
+                ++ showPackageFields pkg
+
+              showPkgStd = showPackageSummaryInfo . mkDisplayInfo srcPkgIndex
+
+              pkgs = InstalledPackageIndex.topologicalOrder installedPkgIndex
+
+          if simpleOutput
+            then putStr $ unlines $ map showPkgSimple pkgs
+            else putStr $ unlines $ map showPkgStd    pkgs
+
+  where
+
+  -- flags:
+  simpleOutput = fromFlag (listSimpleOutput listFlags)
+  fields       = fromNubList $ listFields listFlags
+
+  -- Left if any of the fields are bad, otherwise Right with the show function:
+  possiblyShowPackageFields :: Either String
+                                 (Installed.InstalledPackageInfo -> String)
+  possiblyShowPackageFields =
+    if null fields then
+      Right (\_pkg -> "")
+    else
+      case partitionEithers (map getFieldDescr fields) of
+        ([], fds) -> Right $
+                       \pkg-> '\n' : showFields fds pkg
+
+        (ss, _  ) -> Left $
+                    unlines $
+                      [ "unrecognized package fields: "
+                        ++ unwords ss
+                      , "must use one of these field names:"
+                      ]
+                      ++ map ((" - "++) . fieldName)
+                             Installed.fieldsInstalledPackageInfo
+    where
+    getFieldDescr nm =
+      case [ fd | fd <- Installed.fieldsInstalledPackageInfo
+                , nm == fieldName fd ] of
+         []   -> Left nm
+         fd:_ -> Right fd
+
+    showFields fds = render . nest 2 . ($+$ text "") . ppFields fds
+
+  mkDisplayInfo :: PackageIndex.PackageIndex SourcePackage
```

Comment:
This function looks like something that could be reused in other parts of the file.